src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,1168 @@
+    }
+
+    getPublicKeyFromServer();
+}
+
+QString publicKeyPath(AccountPtr account)
+{
+    return baseDirectory() + account->displayName() + ".pub";
+}
+
+QString privateKeyPath(AccountPtr account)
+{
+    return baseDirectory() + account->displayName() + ".rsa";
+}
+
+bool ClientSideEncryption::hasPrivateKey() const
+{
+    return QFileInfo(privateKeyPath(_account)).exists();
+}
+
+bool ClientSideEncryption::hasPublicKey() const
+{
+    return QFileInfo(publicKeyPath(_account)).exists();
+}
+
+void ClientSideEncryption::generateKeyPair()
+{
+    // AES/GCM/NoPadding,
+    // metadataKeys with RSA/ECB/OAEPWithSHA-256AndMGF1Padding
+    qCInfo(lcCse()) << "No public key, generating a pair.";
+    const int rsaKeyLen = 2048;
+
+    EVP_PKEY *localKeyPair = nullptr;
+
+    // Init RSA
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+
+    if(EVP_PKEY_keygen_init(ctx) <= 0) {
+        qCInfo(lcCse()) << "Couldn't initialize the key generator";
+        return;
+    }
+
+    if(EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, rsaKeyLen) <= 0) {
+        qCInfo(lcCse()) << "Couldn't initialize the key generator bits";
+        return;
+    }
+
+    if(EVP_PKEY_keygen(ctx, &localKeyPair) <= 0) {
+        qCInfo(lcCse()) << "Could not generate the key";
+        return;
+    }
+    EVP_PKEY_CTX_free(ctx);
+    qCInfo(lcCse()) << "Key correctly generated";
+    qCInfo(lcCse()) << "Storing keys locally";
+
+    QDir dir;
+    if (!dir.mkpath(baseDirectory())) {
+        qCInfo(lcCse()) << "Could not create the folder for the keys.";
+        return;
+    }
+
+    auto privKeyPath = privateKeyPath(_account).toLocal8Bit();
+    auto pubKeyPath = publicKeyPath(_account).toLocal8Bit();
+    FILE *privKeyFile = fopen(privKeyPath.constData(), "w");
+    FILE *pubKeyFile = fopen(pubKeyPath.constData(), "w");
+
+    qCInfo(lcCse()) << "Private key filename" << privKeyPath;
+    qCInfo(lcCse()) << "Public key filename" << pubKeyPath;
+
+    //TODO: Verify if the key needs to be stored with a Cipher and Pass.
+    if (!PEM_write_PrivateKey(privKeyFile, localKeyPair, NULL, NULL, 0, 0, NULL)) {
+        qCInfo(lcCse()) << "Could not write the private key to a file.";
+        return;
+    }
+
+    if (!PEM_write_PUBKEY(pubKeyFile, localKeyPair)) {
+        qCInfo(lcCse()) << "Could not write the public key to a file.";
+        return;
+    }
+
+    fclose(privKeyFile);
+    fclose(pubKeyFile);
+
+    generateCSR(localKeyPair);
+
+    //TODO: Send to server.
+    qCInfo(lcCse()) << "Keys generated correctly, sending to server.";
+}
+
+void ClientSideEncryption::generateCSR(EVP_PKEY *keyPair)
+{
+    // OpenSSL expects const char.
+    auto cnArray = _account->davUser().toLocal8Bit();
+    qCInfo(lcCse()) << "Getting the following array for the account Id" << cnArray;
+
+    auto certParams = std::map<const char *, const char*>{
+      {"C", "DE"},
+      {"ST", "Baden-Wuerttemberg"},
+      {"L", "Stuttgart"},
+      {"O","Nextcloud"},
```

Comment:
Are those values above really required? Normally only CN should be a "MUST". I can imagine some customers complaining about this being in the certificate :-)